File: shortest_path/py/bidirectional.py
# Tested
# Verified
def bidirectional(adjList, source, target):
    fronts = [set([source]), set([target])]
    visited = [set([source]), set([target])]
    cnt = [0, 0]
    prev = [{source: None}, {target: None}]
    border = set()
    if source == target:
        border.add(source)
    while all(fronts) and not border: 
        smaller = 0 if len(fronts[0]) < len(fronts[1]) else 1
        children = set()
        cnt[smaller] += 1
        for node in fronts[smaller]:
            for child in adjList[node]:
                if child in fronts[not smaller]:
                    border.add(child)
                if child not in visited[smaller]:
                    visited[smaller].add(node)
                    children.add(child)
                    prev[smaller][child] = node
        fronts[smaller] = children
    for node, parent in prev[1].items():
        prev[0][parent] = node
    return sum(cnt), prev[0]

File: shortest_path/cxx/bidirectional.cxx
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

int bidirectional(std::unordered_map<int, std::vector<int> >& adjList, const int source, const int dest)
{
	std::vector<std::unordered_set<int> > fronts =  {{source}, {dest}};
	std::vector<std::unordered_set<int> > visited = {{source}, {dest}};
	std::unordered_set<int> border;
	while(not fronts[0].empty() and not fronts[1].empty() and border.empty())
	{
		int smaller = fronts[0].size() < fronts[1].size() ? 0 : 1;
		int larger = not smaller;
		std::unordered_set<int> children;
		for(const auto& node: fronts[smaller])
			for(const auto& child: adjList[node])
			{
				if(fronts[larger].find(child) != fronts[larger].end())
					border.insert(child);
				if(visited[smaller].find(child) == visited[smaller].end())
				{
					children.insert(child);
					visited[smaller].insert(child);
				}
			}
		fronts[smaller] = children;
	}
	return 0;
}
